src/build_function.jl

```diff
@@ -351,19 +351,20 @@ end
 get_varnumber(varop, vars::Vector) =  findfirst(x->isequal(x,varop),vars)
 get_varnumber(varop, var) =  isequal(var,varop) ? 0 : nothing
 
-function numbered_expr(O::Symbolic,args...;varordering = args[1],offset = 0,
+buildvarnumbercache(args...) = Dict([isa(arg,AbstractArray) ? el=>(argi,eli) : arg=>(argi,0)
+                                    for (argi,arg) in enumerate(args) for (eli,el) in enumerate(arg)])
+
+function numbered_expr(O::Symbolic,varnumbercache,args...;varordering = args[1],offset = 0,
                        lhsname=gensym("du"),rhsnames=[gensym("MTK") for i in 1:length(args)])
     O = value(O)
     if O isa Sym || isa(operation(O), Sym)
-        for j in 1:length(args)
-            i = get_varnumber(O,args[j])
-            if i !== nothing
-                return i==0 ? :($(rhsnames[j])) : :($(rhsnames[j])[$(i+offset)])
-            end
+        if haskey(varnumbercache,O)
+            (j,i) = varnumbercache[O]
```

Comment:
This could be even faster using `get(varnumbercache, O, nothing)` because `haskey` and `varnumbercache [O]` together needs two hash table lookup.

Reply:
Thanks, done (see commit). However, with the awful scaling behavior from before eliminated, the symbol lookup is really not the bottleneck anymore, the actual expression building takes most of the time. But even with that, one can now build code in acceptable time that easily overburdens gcc. This is actually a separate problem, but not for this PR. ;-)